New module mysql_info - Gather information about MySQL servers

Andersson007 opened the pull request:
##### SUMMARY
New module mysql_info - Gather information about MySQL servers

##### ISSUE TYPE
- New Module Pull Request

##### EXAMPLES
```
- name: Collect all possible information using passwordless root access
  mysql_info:
    login_user: root

- name: Get MySQL version with non-default credentials
  mysql_info:
    login_user: mysuperuser
    login_password: mysuperpass
    filter: version

- name: Collect all info except settings and roles by root
  mysql_info:
    login_user: root
    login_password: rootpass
    filter: "!settings,!roles"

- name: Collect info about databases and version using ~/.my.cnf as a credential file
  become: yes
  mysql_info:
    filter:
    - databases
    - version

- name: Collect info about databases and version using ~alice/.my.cnf as a credential file
  become: yes
  mysql_info:
    config_file: /home/alice/.my.cnf
    filter:
    - databases
```

##### RETURN
Example of return values
```
    "version": {
        "major": 5, 
        "minor": 5, 
        "release": 60
    }
    "databases": {
        "information_schema": {
            "size": 73728
        }, 
        "mysql": {
            "size": 656594
        }, 
    "engines": {
        "CSV": {
            "Comment": "CSV storage engine", 
            "Savepoints": "NO", 
            "Support": "YES", 
            "Transactions": "NO", 
            "XA": "NO"
        }, 
        "InnoDB": {
            "Comment": "Percona-XtraDB, Supports transactions, row-level locking, and foreign keys", 
            "Savepoints": "YES", 
            "Support": "DEFAULT", 
            "Transactions": "YES", 
            "XA": "YES"
        }, 
        "MyISAM": {
            "Comment": "MyISAM storage engine", 
            "Savepoints": "NO", 
            "Support": "YES", 
            "Transactions": "NO", 
            "XA": "NO"
        }, 
    }, 
    "master_status": {
        "Binlog_Do_DB": "", 
        "Binlog_Ignore_DB": "mysql", 
        "File": "mysql-bin.000001", 
        "Position": 769
    }, 
    "roles": {
        "10.44.1.124": {
            "replication_user": {
                "Alter_priv": "N", 
                "Alter_routine_priv": "N", 
                "Create_priv": "N", 
                "Create_routine_priv": "N", 
                "Create_tablespace_priv": "N", 
                "Create_tmp_table_priv": "N", 
                "Create_user_priv": "N", 
                "Create_view_priv": "N", 
                "Delete_priv": "N", 
                "Drop_priv": "N", 
                "Event_priv": "N", 
                "Execute_priv": "N", 
                "File_priv": "N", 
                "Grant_priv": "N", 
                "Index_priv": "N", 
                "Insert_priv": "N", 
                "Lock_tables_priv": "N", 
                "Password": "*E0A25BDCE20352AD0F0EB6706C5FE1F6299D2263", 
                "Process_priv": "N", 
                "References_priv": "N", 
                "Reload_priv": "N", 
                "Repl_client_priv": "N", 
                "Repl_slave_priv": "Y", 
                "Select_priv": "N", 
                "Show_db_priv": "N", 
                "Show_view_priv": "N", 
                "Shutdown_priv": "N", 
                "Super_priv": "N", 
                "Trigger_priv": "N", 
                "Update_priv": "N", 
                "authentication_string": "", 
                "max_connections": 0, 
                "max_questions": 0, 
                "max_updates": 0, 
                "max_user_connections": 0, 
                "plugin": "", 
                "ssl_cipher": "", 
                "ssl_type": "", 
                "x509_issuer": "", 
                "x509_subject": ""
            }
        }, 
    "settings": {
        "innodb_adaptive_flushing": "ON", 
        "innodb_adaptive_flushing_method": "estimate", 
        "innodb_adaptive_hash_index": "ON", 
        "innodb_adaptive_hash_index_partitions": 1, 
        "innodb_additional_mem_pool_size": 8388608, 
        "innodb_autoextend_increment": 8, 
        "innodb_autoinc_lock_mode": 1, 
        "innodb_blocking_buffer_pool_restore": "OFF", 
        "innodb_buffer_pool_instances": 1, 
        "innodb_buffer_pool_populate": "OFF", 
        "innodb_buffer_pool_restore_at_startup": 0, 
        "innodb_buffer_pool_shm_checksum": "ON", 
        "innodb_buffer_pool_shm_key": 0, 
        "innodb_buffer_pool_size": 134217728, 
        "innodb_change_buffering": "all", 
        "innodb_checkpoint_age_target": 0, 
        "innodb_checksums": "ON", 
        "innodb_commit_concurrency": 0, 
        "innodb_concurrency_tickets": 500, 
        "innodb_corrupt_table_action": "assert", 
        "innodb_data_file_path": "ibdata1:10M:autoextend", 
        "innodb_data_home_dir": "", 
        "innodb_dict_size_limit": 0, 
        "innodb_doublewrite": "ON", 
        "innodb_doublewrite_file": "", 
        "innodb_fake_changes": "OFF", 
        "innodb_fast_checksum": "OFF", 
        "innodb_fast_shutdown": 1, 
        "innodb_file_format": "Antelope", 
        "innodb_file_format_check": "ON", 
        "innodb_file_format_max": "Antelope", 
        "innodb_file_per_table": "OFF", 
        "innodb_flush_log_at_trx_commit": 1, 
        "innodb_flush_method": "", 
        "innodb_flush_neighbor_pages": "area", 
        "innodb_force_load_corrupted": "OFF", 
        "innodb_force_recovery": 0, 
        "innodb_ibuf_accel_rate": 100, 
        "innodb_ibuf_active_contract": 1, 
        "innodb_ibuf_max_size": 67092480, 
        "innodb_import_table_from_xtrabackup": 0, 
        "innodb_io_capacity": 200, 
        "innodb_kill_idle_transaction": 0, 
        "innodb_large_prefix": "OFF", 
        "innodb_lazy_drop_table": 0, 
        "innodb_lock_wait_timeout": 50, 
        "innodb_locking_fake_changes": "ON", 
        "innodb_locks_unsafe_for_binlog": "OFF", 
        "innodb_log_block_size": 512, 
        "innodb_log_buffer_size": 8388608, 
        "innodb_log_file_size": 5242880, 
        "innodb_log_files_in_group": 2, 
        "innodb_log_group_home_dir": "./", 
        "innodb_max_bitmap_file_size": 104857600, 
        "innodb_max_changed_pages": 1000000, 
        "innodb_max_dirty_pages_pct": 75, 
        "innodb_max_purge_lag": 0, 
        "innodb_merge_sort_block_size": 1048576, 
        "innodb_mirrored_log_groups": 1, 
        "innodb_old_blocks_pct": 37, 
        "innodb_old_blocks_time": 0, 
        "innodb_open_files": 300, 
        "innodb_page_size": 16384, 
        "innodb_print_all_deadlocks": "OFF", 
        "innodb_purge_batch_size": 20, 
        "innodb_purge_threads": 1, 
        "innodb_random_read_ahead": "OFF", 
        "innodb_read_ahead": "linear", 
        "innodb_read_ahead_threshold": 56, 
        "innodb_read_io_threads": 4, 
        "innodb_recovery_stats": "OFF", 
        "innodb_recovery_update_relay_log": "OFF", 
        "innodb_replication_delay": 0, 
        "innodb_rollback_on_timeout": "OFF", 
        "innodb_rollback_segments": 128, 
        "innodb_show_locks_held": 10, 
        "innodb_show_verbose_locks": 0, 
        "innodb_simulate_comp_failures": 0, 
        "innodb_spin_wait_delay": 6, 
        "innodb_stats_auto_update": 1, 
        "innodb_stats_method": "nulls_equal", 
        "innodb_stats_modified_counter": 0, 
        "innodb_stats_on_metadata": "ON", 
        "innodb_stats_sample_pages": 8, 
        "innodb_stats_traditional": "ON", 
        "innodb_stats_update_need_lock": 1, 
        "innodb_strict_mode": "OFF", 
        "innodb_support_xa": "ON", 
        "innodb_sync_spin_loops": 30, 
        "innodb_table_locks": "ON", 
        "innodb_thread_concurrency": 0, 
        "innodb_thread_concurrency_timer_based": "OFF", 
        "innodb_thread_sleep_delay": 10000, 
        "innodb_track_changed_pages": "OFF", 
        "innodb_use_atomic_writes": "OFF", 
        "innodb_use_fallocate": "OFF", 
        "innodb_use_global_flush_log_at_trx_commit": "ON", 
        "innodb_use_native_aio": "ON", 
        "innodb_use_stacktrace": "OFF", 
        "innodb_use_sys_malloc": "ON", 
        "innodb_use_sys_stats_table": "OFF", 
        "innodb_version": "5.5.59-MariaDB-38.11", 
        "innodb_write_io_threads": 4, 
    }, 
    "slave_hosts": {
        "2": {
            "Host": "", 
            "Master_id": 1, 
            "Port": 3306
        }
    }, 
```

Andersson007 committed:
New module mysql_info: collect info about MySQL server instance

Andersson007 committed:
mysql_info - CI tests

Andersson007 committed:
mysql_info: fixes

Andersson007 committed:
mysql_info: fixes Decimal

Andersson007 committed:
mysql_info: fixes Decimal converters.py

bmalynovytch commented:
-label needs_triage

Andersson007 commented:
@abadger , please, look at the changes in basic.py and converter.py.
Waiting for your answer

Andersson007 commented:
ready_for_review

flowerysong commented on lib/ansible/modules/database/mysql/mysql_info.py:
```diff
@@ -0,0 +1,400 @@
+    # because of it doesn't change anything
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=False
```
This should be supports_check_mode=True; since the module doesn't change anything it doesn't need to do anything special to not change things when run in check mode.

Andersson007 replied:
@flowerysong , thank you for the review! done

bmalynovytch commented on lib/ansible/modules/database/mysql/mysql_info.py:
```diff
@@ -0,0 +1,400 @@
+    # because of it doesn't change anything
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=False
```
```suggestion
        supports_check_mode=True
```
As suggested by @flowerysong

bmalynovytch commented on lib/ansible/module_utils/mysql.py:
```diff
@@ -76,8 +79,30 @@ def mysql_connect(module, login_user=None, login_password=None, config_file='',
+        db_connection = mysql_driver.connect(**config)
+
+    except Exception as e:
+        module.fail_json(msg="unable to connect to database, check login_user and "
```
Can't a network error occur there ? (wrong hostname / IP, network / DNS failure, etc ...)
If so, the error message should be more general and reflect the exception's message rather than adding extra verbosity which doesn't really explain why something went wrong

Andersson007 replied:
Yes, you're absolutely right, thank you. I copy-pasted it from another mysql module and felt that something wrong when testing ;) Now I see, indeed, it should be more general. Done.

Andersson007 committed:
mysql_info: fixed err conn message and check_mode

Andersson007 commented:
ready_for_review

Andersson007 committed:
mysql_info: added decimal.Decimal to lib/ansible/parsing/ajson.py

Andersson007 commented:
@samdoran , regarding that @abadger said:
```
sdoran andersson007_: ^ I'd probably add Decimal to lib/ansible/parsing/ajson.py as well to make an eventual merge of those functions easier.  Other than that, https://github.com/ansible/ansible/pull/55434 looks good to me.
```
I added decimal type to lib/ansible/parsing/ajson.py (in this PR I added it in 3 files: basic.py, converters.py and ajson.py).
Waiting for your feedback

Andersson007 commented:
@bmalynovytch , @flowerysong , please, add shipit if you like this PR

bmalynovytch commented:
@Andersson007 right now (did not have enough time yet to give it a deep eye), I see some whitespace changes without evident value regarding the module addition.
If you feel that the code would benefit from reindentation, I suggest you do it in a different PR, so that people don't get confused about those changes.

Andersson007 commented:
@bmalynovytch , if you mean basic.py, etc, sure, I’d like @samdoran look at this, then we might implement this by another PR and I’ll remove it from this.

Andersson007 commented:
@bmalynovytch , sorry, my find in a fog after the multitask day.
If you mean documentation, we discussed indentation style - that’s OK if it passes sanity tests

bmalynovytch commented:
I was more thinking of:
https://github.com/ansible/ansible/pull/55434/files#diff-574ce10847bde8823e28af27d7cc3839R48
or
https://github.com/ansible/ansible/pull/55434/files#diff-90085fdcec6ed8b273ba885eaee60328R401
😉

Andersson007 commented:
@bmalynovytch , thank you for the explanation;) I’m afraid I’ll need to close and reopen it 10-20 times for test passing if I change it :) I might remove the whitespaces by another PR after merging this ;)

samdoran commented:
@Andersson007 Changes to `basic.py`, `ajson.py`, and `converters.py` look good.

Andersson007 commented:
@samdoran , thank you for reviewing, should I do it by a separate PR?

Andersson007 commented:
We discussed it, I’ll do it by another PR

samdoran commented:
@Andersson007 A separate PR for the core changes would probably be best.

Andersson007 committed:
mysql_info: convert Decimal to float

Andersson007 committed:
mysql_info: fix

Andersson007 committed:
mysql_info: fix

Andersson007 commented:
ready_for_review

Andersson007 committed:
mysql_info: returned the args list as it was

Andersson007 committed:
mysql_info: fix typo

Andersson007 commented:
@bmalynovytch , changed the args string as you like, ready for your review

Andersson007 commented:
Percona guys @michaelcoburn and @Oneiroi , could you please look at this one?
I'm waiting for this module will be merged.
Maybe then I could help with opened issues/PRs/new modules for MySQL.

cmfrtl commented:
shipit

Andersson007 commented:
@txracs , could you please look at this ?

Andersson007 commented:
@tcraxs , could you please look at this ?

Andersson007 commented:
@bmalynovytch ping, lets do something with it. I'll review at least 2 MySQL PRs (on your choice) after this one is merged ;)

bmalynovytch commented:
@Andersson007 Hahaha ! 😄
I'll try to give it an eye ASAP 👍

bmalynovytch commented:
shipit

Andersson007 commented:
@bmalynovytch , thank you for reviewing, feel free to ping me directly by aaklychkov at mail dot ru

Andersson007 committed:
mysql_info: fixes

Andersson007 committed:
mysql_info: removed aliases for login_db

Andersson007 committed:
mysql_info: changed RETURN condition info

Andersson007 committed:
mysql_info: fixed doc

ffdybuster commented on lib/ansible/modules/database/mysql/mysql_info.py:
```diff
@@ -0,0 +1,403 @@
+            for line in res:
+                host = line['Host']
+                user = line['User']
+                self.info['roles'][host] = {user: {}}
```
If you have more than one user per host, only the last one will end up in the results.

Andersson007 replied:
thank you! fixed

ffdybuster commented on lib/ansible/modules/database/mysql/mysql_info.py:
```diff
@@ -0,0 +1,403 @@
+                host = line['Master_Host']
+                port = line['Master_Port']
+                user = line['Master_User']
+                self.info['slave_status'][host] = {port: {user: {}}}
```
Same problem here.

Andersson007 commented:
@ffdybuster , thank you! Going to check it ASAP

Andersson007 committed:
mysql_info: fixed role's parsing

Andersson007 committed:
mysql_info: fixed role's parsing

Andersson007 committed:
mysql_info: fixes

Andersson007 committed:
mysql_info: fixed integration tests

Andersson007 commented:
ready_for_review

felixfontein reviewed (approved):
LGTM

Andersson007 commented:
@felixfontein , thank you for reviewing! This module became significantly better and we avoided several bugs, thanks!

felixfontein commented:
@Andersson007 thanks for adding this!
Everyone else, thanks for reviewing!

Andersson007 commented:
thanks all!
